### PR TITLE
Add NLL tests for #46557 and #38899

### LINKED ISSUE #38899: borrowed referent of a `&T` sometimes incorrectly allowed
@jorendorf asks on [the users forum](https://users.rust-lang.org/t/why-are-reads-sometimes-tolerated-during-mut-borrows-and-sometimes-not/8774) about a curious discrepancy around fields. It seems that implicit borrows sometimes seem to get overlooked in the borrow checker.  This seems like a kind of bad bug, though it's exact scope is unclear until we investigate a bit more. 

Here is a variant of @jorendorf's example which is pretty clearly wrong. Here, the `block` variable is mutably borrowed into `x`, so it should not be accessible via `let p`:

```rust
#![allow(dead_code)]
pub struct Block<'a> {
    current: &'a u8,
    unrelated: &'a u8,
}

fn bump<'a>(mut block: &mut Block<'a>) {
    let x = &mut block;
    let p: &'a u8 = &*block.current;
	// (use `x` and `p` so enabling NLL doesn't assign overly short lifetimes)
	drop(x);
	drop(p);
}

fn main() {}
```

I'm guessing that the problem has to do with the logic around borrowing the referent of an `&T` (in this case, we are borrowing `*block.current`). In particular, we deem that to be "safe" for the scope of `'a` because the data is independently guaranteed to be valid that long (this is reasonable). But we still need to validate that `block.current` can be (instantaneously) read. It seems we are not doing that. But this is all a hypothesis: I've not dug into the code to validate it.

### LINKED ISSUE #46557: Compiler accepts return mut ref to local var on no longer valid stackframe
Spotted on reddit: https://www.reddit.com/r/rust/comments/7i64sy/safe_function_returns_static_mut/  
credit to [jDomantas](https://www.reddit.com/user/jDomantas)

I tried this code: [playground](https://play.rust-lang.org/?gist=5e19e59e8b64488171b76bfbf0383b06&version=stable)

	fn gimme_static_mut() -> &'static mut u32 {
	    let ref mut x = 1234543;
	    x
	}

	fn main() {
	    let x = gimme_static_mut();
	    let y = gimme_static_mut();
	    *y = 42;
	    let a = *x;
	    println!("a: {}", a);
	}

This is accepted by the compiler and prints 42.

I expected to see this happen: The program should be rejected by the compiler with a lifetime error

Instead this happened: The program was accepted and miscompiled badly

The assembly code clearly demonstrates the problem, a local variable is created for the literal, its address is taken and returned:

	playground::gimme_static_mut:
	.Lfunc_begin2:
		push	rbp
	.Lcfi6:
	.Lcfi7:
		mov	rbp, rsp
	.Lcfi8:
		sub	rsp, 16
		lea	rax, [rbp - 4]
	.Ltmp6:
		mov	dword ptr [rbp - 4], 1234543
		mov	qword ptr [rbp - 16], rax
		mov	rax, qword ptr [rbp - 16]
	.Ltmp7:
		add	rsp, 16
		pop	rbp
		ret
	.Ltmp8:
	.Lfunc_end2:

## Meta

Latest version on playground

### LINKED ISSUE #47366: tracking issue for bugs fixed by the MIR borrow checker or NLL
Part of: https://github.com/rust-lang/rust/issues/43234

The MIR-based borrow checker used by NLL fixes a number of bugs in the existing AST-based borrow checker. Our plan for these bugs is to fix them by transitioning to the MIR-based borrow checker. In the meantime, we wish to add tests that show that the bugs are fixed with NLL, and then close the bugs (so they don't clutter up the issue tracker).  This issue tracks the set of such bugs and gives instructions for how to proceed.

## Bugs fixed by MIR

All these bugs are **confirmed as fixed**; the checkmark indicates that a PR with a test is enqueued or landed.

- [x] https://github.com/rust-lang/rust/issues/46557
- [x] https://github.com/rust-lang/rust/issues/38899
- [x] #45045 (using example from [this comment](https://github.com/rust-lang/rust/issues/45045#issuecomment-356953507))
- [x] https://github.com/rust-lang/rust/issues/46314
- [x] #47053  (PR https://github.com/rust-lang/rust/pull/47425)
- [x] https://github.com/rust-lang/rust/issues/47366 (examples from [this comment])
- [x] https://github.com/rust-lang/rust/issues/29149 (fix in https://github.com/rust-lang/rust/pull/47877)
- [x] #16223 
- [x] #48238 (and specifically [this reduction](https://github.com/rust-lang/rust/issues/48238#issuecomment-366048269))
- [x] https://github.com/rust-lang/rust/issues/45157
- [x] https://github.com/rust-lang/rust/issues/46314 (hygiene interaction)
- [x] https://github.com/rust-lang/rust/issues/27868 -- evaluation order and operators
- [x] https://github.com/rust-lang/rust/issues/16223 -- incorrect partially moved value error check

## Procedure

- First, if you wish, add yourself to the [list of participants for the WG-compiler-nll working group](https://docs.google.com/spreadsheets/d/1vymoSTIzepS6PVidHJWsMOVl12aSXixMev0LBLKLykY/edit?usp=sharing).
    - This will help us to find people who might like to help later!
- Create a regression test in `src/test/ui/nll/`
    - The name can be `issue-XXX.rs`, though giving a more descriptive name is also good
    - The file should include a comment saying `// Regression test for #XXX` however to ease searching later
- This test should include the error from the bug along with the following header material:
    - Rust license
    - `#![feature(nll)]` at the top
    - comments like `//~ ERROR` for the errors, or whatever is needed for test harness to pass
- Run `./x.py test --stage 1 src/test/ui --test-args nll`, this will test all the nll tests (including yours)
- Presuming that passes, open a PR! Otherwise, stop in at the [WG-compiler-nll gitter channel](https://gitter.im/rust-impl-period/WG-compiler-nll)
    - Put `r? @nikomatsakis` in the PR summary

### PR DESCRIPTION
**Pull request by chrisvittal**
This adapts the sample code from the two issues into test code.

Closes #46557 
Closes #38899 

r? @nikomatsakis

### CODE REVIEW

**Comment by chrisvittal** [comment]
The mentoring instructions were unclear here. It seems that we want these tests to fail compilation as they both unsoundly passed AST borrowck.

**Comment by pnkfelix** [comment]
@chrisvittal When you said the instructions were unclear, were you referring to the step that said that the tests should include the comment `// must-compile-successfully` ? (Because I agree: the two cases that @nikomatsakis listed are ones where the whole point is that they start *failing* to compile under `#![feature(nll)]`

**Comment by pnkfelix** [comment]
Also I think the intention is that this PR should be closing  #46557  and #38899 . I don't know if it is premature to close #47366 as well, though I guess we can always reopen it if we find other MIR issues...

**Comment by nikomatsakis** [comment]
That was a mistake on my part. We *expect* errors!

File: src/test/ui/nll/borrowed-referent-issue-38899.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Regression test for issue #38899
+
+// compile-flags:-Znll -Zborrowck=mir -Zverbose
```
Comment (nikomatsakis) [inline src/test/ui/nll/borrowed-referent-issue-38899.rs]: can we remove the `compile-flags`? They should not be necessary. `feature(nll)` should suffice.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
(also remove flags from other test)

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**